File: cpp/ransom-note.cpp
/*
Time complexity: O(m + n), since we traverse the magazine once to preprocess
                 it, and traverse the ransom note once to remove frequencies
                 from preprocessing.
Space complexity: O(n), since we store an occurrence of each character in the
                  magazine after we preprocess it.
*/

class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        const int magazineLength = magazine.length();
        const int ransomNoteLength = ransomNote.length();

        // preprocess the magazine
        preprocess(magazine, magazineLength);
        return isValid(ransomNote, ransomNoteLength);
    }

private:
    unordered_map<char, int> magMap;
    void preprocess(const string& str, const int size) {
        for (int ch = 0; ch < size; ++ch) {
            magMap[str[ch]] += 1;
        }
        
    }

    bool isValid(const string& str, const int size) {
        for (int ch = 0; ch < size; ++ch) {
            magMap[str[ch]] -= 1;
            if (magMap[str[ch]] < 0) {
                return false;
            }
        }
        return true;
    }
};



class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        constexpr int bucketCount = 26;
        int chars[26] = {0};
        const int magazineLength = magazine.length();
        const int ransomNoteLength = ransomNote.length();

        // All letters are lowercase, substract 97 from their ASCIIto index them from 0-26
        for (int i = 0; i < magazineLength; ++i) {
            ++(chars[magazine[i] - 97]);
        }

        for (int i = 0; i < ransomNoteLength; ++i) {
            --(chars[ransomNote[i] - 97]);
        }

        for (int i = 0; i < bucketCount; ++i) {
            if (chars[i] < 0) {
                return false;
            }
        }
        return true;
    }
};

File: py/ransom-note.py
from collections import defaultdict

# O(n) time and space to preprocess the magazine
# O(m) time to check if each character occurence is inside of ransomNote
# Time complexity is O(n + m) to traverse the ransomNote and magazine each once
# Space complexity is O(n) to preprocess the magazine

class Solution:
    def __init__(self):
        # initializes a dict with all zeroes
        self.magazineDict = defaultdict(int)

    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        self.preprocessMagazine(magazine)
        return self.checkRansomNote(ransomNote)


    def preprocessMagazine(self, mag: str) -> None:
        for char in mag:
            self.magazineDict[char] += 1

    def checkRansomNote(self, note: str) -> bool:
        for char in note:
            self.magazineDict[char] -= 1
            if self.magazineDict[char] < 0:
                return False
        return True

